test/filters/test_crop_hull.cpp

```diff
@@ -0,0 +1,382 @@
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *
+ * $Id: test_filters.cpp 7683 2012-10-23 02:49:03Z rusu $
+ *
+ */
+
+#include <pcl/test/gtest.h>
+#include <pcl/pcl_tests.h>
+
+#include <random>
+
+#include <pcl/point_types.h>
+#include <pcl/surface/convex_hull.h>
+#include <pcl/common/common.h>
+#include <pcl/filters/crop_hull.h>
+
+
+namespace
+{
+  struct Checker
+  {
+    pcl::CropHull<pcl::PointXYZ> cropHullFilter;
+
+    Checker(pcl::ConvexHull<pcl::PointXYZ> & hull, int dim)
+      : cropHullFilter(true)
+    {
+      pcl::PointCloud<pcl::PointXYZ>::Ptr hullCloudPtr(new pcl::PointCloud<pcl::PointXYZ>);
+      std::vector<pcl::Vertices> hullPolygons;
+      hull.reconstruct(*hullCloudPtr, hullPolygons);
+
+      cropHullFilter.setHullIndices(hullPolygons);
+      cropHullFilter.setHullCloud(hullCloudPtr);
+      cropHullFilter.setDim(dim);
+      cropHullFilter.setCropOutside(true);
+    }
+
+    void check(
+        pcl::Indices const & expectedFilteredIndices,
+        pcl::PointCloud<pcl::PointXYZ>::ConstPtr inputCloud)
+    {
+      std::vector<bool> expectedFilteredMask(inputCloud->size(), false);
+      pcl::Indices expectedRemovedIndices;
+      pcl::PointCloud<pcl::PointXYZ> expectedCloud;
+      pcl::copyPointCloud(*inputCloud, expectedFilteredIndices, expectedCloud);
+      for (pcl::index_t idx : expectedFilteredIndices) {
+        expectedFilteredMask[idx] = true;
+      }
+      for (size_t i = 0; i < inputCloud->size(); ++i) {
+        if (!expectedFilteredMask[i]) {
+          expectedRemovedIndices.push_back(i);
+        }
+      }
+
+      cropHullFilter.setInputCloud(inputCloud);
+
+      pcl::Indices filteredIndices;
+      cropHullFilter.filter(filteredIndices);
+      pcl::test::EXPECT_EQ_VECTORS(expectedFilteredIndices, filteredIndices);
+      pcl::test::EXPECT_EQ_VECTORS(expectedRemovedIndices, *cropHullFilter.getRemovedIndices());
+      // check negative filter functionality
+      {
+        cropHullFilter.setNegative(true);
+        cropHullFilter.filter(filteredIndices);
+        pcl::test::EXPECT_EQ_VECTORS(expectedRemovedIndices, filteredIndices);
+        pcl::test::EXPECT_EQ_VECTORS(expectedFilteredIndices, *cropHullFilter.getRemovedIndices());
+        cropHullFilter.setNegative(false);
+      }
+      // check cropOutside functionality
+      {
+        cropHullFilter.setCropOutside(false);
+        cropHullFilter.filter(filteredIndices);
+        pcl::test::EXPECT_EQ_VECTORS(expectedRemovedIndices, filteredIndices);
+        pcl::test::EXPECT_EQ_VECTORS(expectedFilteredIndices, *cropHullFilter.getRemovedIndices());
+        cropHullFilter.setCropOutside(true);
+      }
```

Comment:
(1) The core code of L89-L108 checks the filtered and removed indices and can be extracted into a call from inside this function. Same with L110-L134 which check the filtered cloud

```cpp
void some_test(data) {
  pcl::Indices filteredIndices;  // variable setup for computing actual data
  some_filter.filter(filteredIndices);  // no modification of some_filter.
  run_tests  // any expected data to check is passed to the function
}
```